### PR TITLE
Point changes in customization and profile tabs to users_settings table

### PR DESCRIPTION
**Pull request by msarit**
## Description
This PR serves as a "base" for the PRs I am creating that point changes to UsersSetting and UsersNotificationSetting attributes from `users` table to `users_settings` and `users_notification_settings` tables.
I am doing it this way to aid PR reviewers as much as possible.
Please review the Action plan below, and contact Arit if you need more clarification. Thank you for tackling this code review! ❤️ 

## ACTION PLAN (for [Step 5](https://github.com/forem/rfcs/blob/2dd7c4d32be2d72aca05389d53f614465ecff570/text/0000-user-settings.md) of the UsersSettings RFC)
1. Complete the implementation of the following tasks:
- Point changes on `brand_color1`, `brand_color2`, `config_theme`, `config_font`, `config_navbar`, `display_announcements`, `display_email_on_profile`, `display_sponsors`, `editor_version`, `experience_level`, `permit_adjacent_sponsors` from `users` table to `users_settings` table. [PR 13904](https://github.com/forem/forem/pull/13904) [this present PR] does this. ✅ 
- Point changes on `feed_fetched_at`, `feed_mark_canonical`, `feed_referential_link`, `feed_url` attributes from `users` table to `users_settings` table. [PR 13914](https://github.com/forem/forem/pull/13914). ✅ 
- Point changes on `inbox_type` and `inbox_guidelines` attributes from `users` table to `users_settings` table. [PR 13911](https://github.com/forem/forem/pull/13911). ✅ 
- Point changes on [all notification attributes](https://github.com/forem/forem/blob/ab6969b25c0dac27fe647566f7619ed220d70d4c/app/models/user.rb#L118) from `users` table to `users_settings` table. [PR 13910](https://github.com/forem/forem/pull/13910). ✅ 
- Delete the sync code that keeps the `users_settings` and `users_notification_settings` tables up-to-date with the values of all these attributes in `users` table. ✅   
2. Merge the PRs ~13910~, ~13911~ and ~13914~ into this present PR. ✅
3. Write new (and update existing) specs for the new functionality. _Note: I am choosing to write the specs altogether in order to avoid inevitable duplication of effort._ 😅 
4. Merge the consolidated PR 13904 into `main` after all CI tests pass.

## TO-DO list for this PR
- [x] Update relevant attributes in the `customization` and `profile` tabs to use the new `users_settings` table 
- [x] Ensure that the themes, fonts, etc. are applied when page is refreshed
- [X] Deleting a User should delete their associated records in `users_settings` and `users_notification_settings` tables
- [X] Go through the TO-DO's on the application 
- [X] Check for other instances where these fields are being used
- [ ] Fix all the tests and write new ones for the new controllers and concerns

## What type of PR is this? (check all applicable)

- [X] Refactor
- [ ] Feature
- [ ] Bug Fix
- [X] Optimization
- [ ] Documentation Update

## QA Instructions, Screenshots, Recordings

1. While logged in to localhost, change the following settings and ensure they reflect in the app:
a. In the `settings/profile` tab, check/uncheck `Display email on profile` then check your profile to see if changes reflect.
b. Also within `settings/profile`, scroll down to `Branding` and choose two colors. Save, refresh and then check your profile to ensure that your banner color reflects the color you chose.
c. In the `settings/customization` tab, choose a combo of Site Theme, Base Reading Font and Site Navbar. Save, refresh, then ensure that these selections reflect in the app.
d. Also within `settings/customization`, switch your Editor Version, save, then Create a Post. You should see the editor version you chose.
e. To QA Display Announcements and Sponsors, go to `localhost:3000/admin` as an admin and create Announcements and Sponsors under the Advanced tab. Bear in mind that you can have just one sponsor per org_id. Next, within `settings/customization`, toggle Display Announcements and Sponsors and ensure these settings reflect in the app.
f. In the `settings/extensions` tab, scroll to the Publishing ... from RSS section. Enter `https://dev.to/feed` as your RSS Feed URL. Save, then refresh. A `Fetch Feed Now` button should now be visible within this section (if you previously had a Feed URL entered, this button was already visible). Click the button, then refresh your page. The Last Fetched value should be your current time.
2. For all other attributes (especially those in `notification_settings`), toggle them in the app, then in Rails Console, check that your user has the updated values in their `users_setting` and `users_notification_setting` records.

## Added tests?

- [ ] Yes
- [ ] No, and this is why:
- [ ] I need help with writing tests
- [X] Not yet

## [Forem core team only] How will this change be communicated?

- [X] I will share this change internally with the appropriate teams

### CODE REVIEW

File: app/decorators/user_decorator.rb
Comment (msarit) [inline app/decorators/user_decorator.rb:54]: These changes are necessary to pull the customization values from `users_settings` and not `users`.

File: app/decorators/user_decorator.rb
Comment (msarit) [inline app/decorators/user_decorator.rb:63]: same here as above

File: app/views/users/_customization.html.erb
Comment (msarit) [inline app/views/users/_customization.html.erb:24]: The behavior is thus:
1. User selects `default` and saves.
2. The option selected is `sans_serif` which is the default font stored in `Settings::UserExperience.default_font`.

File: app/models/user.rb
Comment (jgaskins) [inline app/models/user.rb:317]: Should we delete this line instead of commenting it out? We can always get it back since it's in git.

```suggestion
```

File: app/models/users/setting.rb
Comment (jgaskins) [inline app/models/users/setting.rb:5]: Is this being used for cache busting?

**Review by jgaskins — APPROVED** [review]

File: app/controllers/users/settings_controller.rb
Comment (msarit) [reply app/controllers/users/settings_controller.rb:30]: To be honest, that wasn't actually me 😅  but I'll accept the thanks! 😊

File: app/models/users/setting.rb
Comment (msarit) [reply app/models/users/setting.rb:5]: (I had to think about your question for a few minutes - cos I was like "what is cache busting?" 🤔 )

Yes! yes we are. When certain settings (eg `config_theme`) are changed, in order for a refresh to apply the changes, the cached user [here](https://github.com/forem/forem/blob/8d632d14478a3ca14fb09d4b8fd6f567e03945cf/app/controllers/async_info_controller.rb#L76) needs to be updated.

I hope that makes sense; I'm explaining it as best I understand.

Comment (jgaskins) [reply app/models/users/setting.rb:5]: That makes perfect sense. Knowing that we're caching content for user settings makes this make a lot of sense. Thanks for the link to where it's used, too! 🎉

**Comment by joshpuetz** [comment]
Swapping out @Ridhwana for my review as she has more experience with this area!

**Comment by Ridhwana** [comment]
> Swapping out @Ridhwana for my review as she has more experience with this area!

Thanks @joshpuetz I will go through the code and I'm always happy to review, however I try and encourage other team members to review the User Settings PR's since I tend to pair with Arit on some of the initial the code and it doesn't make too much sense for me to review it. Thus, if other team members review then it allows for a more diverse perspective on the code.

**Comment by joshpuetz** [comment]
> Thanks @joshpuetz I'll go through the code and I'm always happy to review, however I try and encourage other team members to review the User Settings PR's since I tend to pair with Arit on some of the initial the code and it doesn't make too much sense for me to review it. Thus, if other team members review then it allows for a more diverse perspective on the code.

Whoops, sorry, didn't know that was an explicit goal!

I'm still keeping myself off of review @msarit: I don't have the bandwidth to respond to this!

**Review by Zhao-Andy — DISMISSED** [review]
Overall the code looks good!

I noticed that when I update any settings in `/settings/customization` that while the database updated, I didn't see any changes on the front end. I think maybe there's some missing data that's not being carried over properly.

For example, changing my theme persisted in the form, but not in the UI.

**Comment by msarit** [comment]
@Zhao-Andy Fixed it, Andy 😅

**Review by Zhao-Andy — COMMENTED** [review]
Customization settings updated properly :) Specs are failing though so I'll wait for approval. Feel free to request me again for review.

File: app/decorators/user_decorator.rb
Comment (citizen428) [reply app/decorators/user_decorator.rb:54]: @msarit This decorator also needs changes in the `enriched_colors` method. This currently uses `bg_color_hex` and `text_color_hex` which only work because of this block in `user.rb`:

```ruby
      # Getters and setters for mapped profile attributes
      Profile::MAPPED_ATTRIBUTES.each do |profile_attribute, user_attribute|
        define_method(user_attribute) { profile&.public_send(profile_attribute) }
        define_method("#{user_attribute}=") do |value|
          profile&.public_send("#{profile_attribute}=", value)
        end
      end
```

Instead, this should be using the `brand_color1` and `brand_color2 values from the user's settings.

**Review by Ridhwana — CHANGES_REQUESTED** [review]
@msarit:

I pulled down the branch and tested manually, I haven't completed a code review as per say but rather just clicked around the UI, here are some of the things I picked up up from the QA manually:

To setup my QA I did the ffg: 

- I destroyed my database.
- Created a new database and migrated
- setup the first user with the Create Account button
- Navigated to the /settings page and tested the profile page as a start:


1. Onboarding seems to be broken 
2. I got the ffg error when navigating to `/settings` -  ```undefined method display_email_on_profile' for nil:NilClass```
This is because there is no users_setting record for that user. I explain this alongside your commit in the code.
3. Profile Page: `brand_color_1` and `brand_color_2` are not propagating their changes to the user page i.e. `localhost:3000/<username>`
4. Customization Page: When changing the editor version to `Basic Markdown`, it does not propogate the change to the actual editor on `localhost:3000/new`
5. Can you double check that your sponsors section and Announcements sections are propagating through the app as well - I was pressed for time so I didn't check this out.

It's useful to pull up the pages on https://canary2.forem.wtf/settings and see what their behaviour should be if you're not sure. 


Also, When you get a chance, can you please add some QA instructions to the PR description? From the title and code it seems that I should just test the Profile and Customization tabs but it would be useful to add that in the QA section as well.

File: app/models/user.rb
Comment (msarit) [reply app/models/user.rb:353]: > Since we remove the sync_users_settings_table method completely, there is no users_setting record being associated when a user is created.

🤦🏾‍♀️  🤦🏾‍♀️  🤦🏾‍♀️  🤦🏾‍♀️  Thank you 🙏🏾

File: app/controllers/admin/invitations_controller.rb
Comment (msarit) [inline app/controllers/admin/invitations_controller.rb:27]: `editor_version` is no longer called on `User` but on `UsersSetting`, and defaults to 'v2' in `UsersSetting`. so I thought to remove it from this `invite!` action.

File: app/controllers/registrations_controller.rb
Comment (msarit) [inline app/controllers/registrations_controller.rb:83]: In this case, `resource` is the user, which is why I call `setting`. The `&` is for incase the `setting` record does not exist at this point; the `editor_version` defaults to 'v2' regardless

File: app/decorators/user_decorator.rb
Comment (msarit) [reply app/decorators/user_decorator.rb:54]: Done!

File: app/decorators/user_decorator.rb
Comment (msarit) [inline app/decorators/user_decorator.rb:114]: These methods are to load these 2 `UsersSetting` values into the `UserData` object.

File: app/models/profile.rb
Comment (msarit) [inline app/models/profile.rb:27]: @citizen428 Please note 🙏🏾

File: app/models/user.rb
Comment (msarit) [inline app/models/user.rb:227]: @Ridhwana @rhymes @citizen428 I'm not clear on these two comments; it may be coding fatigue on my part 😓

File: app/models/users/setting.rb
Comment (msarit) [inline app/models/users/setting.rb:27]: This method used to live in `user.rb`; moved it over to `setting.rb` now that `feed-url` is in `users_settings`

File: db/schema.rb
Comment (msarit) [inline db/schema.rb:1436]: I cant think HOW this got deleted 😖  I'm open to suggestions on correcting

Comment (citizen428) [reply db/schema.rb:1436]: For the purpose of fixing this PR, you can just add the line back in the same spot.

File: app/controllers/users/notification_settings_controller.rb
Comment (citizen428) [inline app/controllers/users/notification_settings_controller.rb:28]: I don't think the `notice` variable is necessary, I would combine the two lines into

```ruby
flash[:settings_notice] = "Your notification settings have been updated."
```

This may well be a holdover from old code you copy/pasted but since we're here already maybe we can clean up a little bit.

File: app/controllers/users/settings_controller.rb
Comment (citizen428) [inline app/controllers/users/settings_controller.rb:27]: Again, I don't think this variable adds a lot but if you want to keep it, please move it closer to where it's actually used, it's a bit confusing up here (i.e. I was looking if it gets modified along the way).

File: app/models/user.rb
Comment (rhymes) [inline app/models/user.rb:575]: you can use `create_setting` and `create_notification_setting`, they are special methods [added to Rails by `has_one`](https://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_one)

File: app/models/user.rb
Comment (rhymes) [inline app/models/user.rb:67]: I think you should also hide all the columns that have been moved over, this way if any of the tests exercise some code that are still trying to use those columns, they will break.

See https://api.rubyonrails.org/classes/ActiveRecord/ModelSchema/ClassMethods.html#method-i-ignored_columns-3D

You can find an example here

https://github.com/forem/forem/blob/ff3be35d6e153b113493b48f75a96297d7cdaa26/app/models/podcast_episode.rb#L1-L4

File: config/routes.rb
Comment (rhymes) [inline config/routes.rb:140]: blocking: same suggestion I made in the notification setting PR. Only the current user can update their own setting, so you can use a singular resource here as well

Let me know if that's not the case!

File: db/schema.rb
Comment (rhymes) [inline db/schema.rb:1486]: this be removed as well

**Review by rhymes — COMMENTED** [review]
This is one big PR, thanks for staying on top of this!

I left some notes :)

File: app/models/user.rb
Comment (msarit) [reply app/models/user.rb:67]: perfect point thank you

File: app/models/user.rb
Comment (msarit) [reply app/models/user.rb:575]: Noice!

File: app/models/user.rb
Comment (msarit) [reply app/models/user.rb:481]: I will look into this; I tried yesterday but my brain was fried 🍳

File: app/controllers/profiles_controller.rb
Comment (msarit) [reply app/controllers/profiles_controller.rb:18]: oh it's not tiny at all - good catch 😅

File: app/decorators/user_decorator.rb
Comment (msarit) [reply app/decorators/user_decorator.rb:114]: Noiceeeee! 😎

File: app/models/profile.rb
Comment (citizen428) [reply app/models/profile.rb:27]: Yes, you can remove it, it will no longer be necessary. Thank you!

**Comment by citizen428** [comment]
@msarit When I go to `/settings/customization` I see the following error:

image

This can be fixed by adding the `url` to the form: 

```diff
diff --git a/app/views/users/_customization.html.erb b/app/views/users/_customization.html.erb
index 68cd47d48..fbf30ef99 100644
--- a/app/views/users/_customization.html.erb
+++ b/app/views/users/_customization.html.erb
@@ -1,6 +1,8 @@
 <%= javascript_packs_with_chunks_tag "stickySaveFooter", defer: true %>

-<%= form_for @users_setting, html: { id: "ux-customization-form", class: "sticky-footer-form" } do |f| %>
+<%= form_for @users_setting,
+    url: users_settings_path,
+    html: { id: "ux-customization-form", class: "sticky-footer-form" } do |f| %>
   <%= f.hidden_field :user_id, value: @users_setting.user_id, id: nil %>
   <%= f.hidden_field :tab, value: @tab, id: nil %>
```

If we fix, the form will work, but updates still fail:

image

This is because the controller tries to update `current_user.users_setting` but the association is called `setting`:

```ruby
  has_one :setting, class_name: "Users::Setting", dependent: :destroy
```

Updating this takes us one step further but then fails on the next line where you try to access `users_setting` which doesn't exist as a variable in this scope.

So I guess this controller still needs a bit of work.

**Comment by citizen428** [comment]
When trying to update values on http://localhost:3000/settings/profile, I end up with an error too:

image

This happens because `Profile::Update` divides its updates between the `Profile` and `User` model. For the latter, this means we're running into the following issue:

```ruby
  before_validation :set_config_input

  def set_config_input
    self.config_theme = config_theme&.tr(" ", "_")
    self.config_font = config_font&.tr(" ", "_")
    self.config_navbar = config_navbar&.tr(" ", "_")
  end
```

These `config_*` attributes no longer exist on `User` but are on `Users::Setting` now so this should probably move over. 

After moving the above from `User` to `Users::Setting` we encounter another problem:

image

Please let me know if you get stuck on updating `Profiles::Update` and I'll help you with that.

**Comment by msarit** [comment]
@Ridhwana @citizen428 @rhymes Thank you so much for tirelessly reviewing my work in this PR.

I'm taking a few days break from this work, to give my mind a mental rest, and recover from coding fatigue.

I'll be back Tuesday 22nd or Wednesday 23rd June.

**Comment by Zhao-Andy** [comment]
Going to remove myself from review but feel free to request me again :)

File: app/models/user.rb
Comment (Ridhwana) [reply app/models/user.rb]: TIL!

File: app/models/profile.rb
Comment (msarit) [reply app/models/profile.rb:27]: 👍🏾

File: app/services/email_digest.rb
Comment (msarit) [inline app/services/email_digest.rb:28]: I need help reworking this query; `email_digest_periodic` now lives in `users_notification_setting`, not `users`. My attempts to call `.joins(:users_notification_settings)` on `User` did not work. 😞  I really need to strengthen my DB skills.

File: app/services/notifications/moderation.rb
Comment (msarit) [inline app/services/notifications/moderation.rb:7]: I need help reworking this query; `mod_roundrobin_notifications` now lives in `users_notification_setting`, not `users`.

File: app/services/notifications/new_comment/send.rb
Comment (msarit) [inline app/services/notifications/new_comment/send.rb:38]: I need help reworking this query; `mobile_comment_notifications` now lives in `users_notification_setting`, not `users`.

File: app/services/email_digest.rb
Comment (citizen428) [reply app/services/email_digest.rb:28]: @msartit You need to join on the association name you specified in `User`:

```ruby
User
  .registered
  .joins(:notification_setting)
  .where(notification_setting: { email_digest_periodic: true })
  .where.not(email: "")
```

Note that I specified explicitly that we want to use the `email_digest_periodic` column of the `users_notification_settings` table since it still exists on both tables.

File: app/services/notifications/moderation.rb
Comment (citizen428) [reply app/services/notifications/moderation.rb:7]: Same as above, re explicitly specifying where to find the column:

```ruby
User
  .with_role(:trusted)
  .joins(:notification_setting)
  .where("last_moderation_notification < ?", 3.days.ago)
  .where(notification_setting: { mod_roundrobin_notifications: true })
```

File: app/services/notifications/new_comment/send.rb
Comment (citizen428) [reply app/services/notifications/new_comment/send.rb:38]: This is a variation on the same pattern:

```ruby
User
  .joins(:notification_setting)
  .where(notification_setting: { mobile_comment_notifications: true })
  .ids
```

**Comment by msarit** [comment]
List of Broken Features/Functionality:
- User Profile Images are broken (Org profile images are fine)
- In `settings/extension` the `Fetch Feed` functionality is wonky - sometimes it fetches the feed, and sometimes it does not

**Comment by mstruve** [comment]
closing in favor of https://github.com/forem/forem/pull/14121

**Pull request closed without merging**
